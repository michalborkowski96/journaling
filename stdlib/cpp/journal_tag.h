template<typename T>
class Type_JournalTag_;

template<typename T>
class Type_JournalTag {
public:
	🍆::WeakObject<Type_JournalTag> this_ptr;
	🍆::StrongObject<Type_JournalTag> get_this() {
		return 🍆::scast<Type_JournalTag>(this_ptr);
	}
	🍆::StrongObject<T> object;
	🍆::Journal::Tag tag;
	virtual ~Type_JournalTag() = default;

	Type_JournalTag(WeakObject<Type_JournalTag> this_ptr, 🍆::StrongObject<T> var_t) : this_ptr(this_ptr), object(var_t), tag(var_t->journal.get_tag()){
	}

	Type_JournalTag() = delete;

	Type_JournalTag(const Type_JournalTag&) = delete;

	Type_JournalTag(Type_JournalTag&&) = delete;

	void basefun_bring_back() {
		object->journal.goto_tag(tag);
	}

	virtual void privfun_bring_back() {
		(basefun_bring_back());
	}
	virtual void fun_bring_back() {
		(basefun_bring_back());
	}
	void basefun_t_(🍆::StrongObject<Type_JournalTag_<T>> var_ttt, 🍆::StrongObject<T> var_tt) {
		((var_ttt)->fun_t((var_tt)));
	}

	virtual void privfun_t_(🍆::StrongObject<Type_JournalTag_<T>> var_ttt, 🍆::StrongObject<T> var_tt) {
		(basefun_t_(var_ttt, var_tt));
	}
	virtual void fun_t_(🍆::StrongObject<Type_JournalTag_<T>> var_ttt, 🍆::StrongObject<T> var_tt) {
		(basefun_t_(var_ttt, var_tt));
	}
};

template<>
class Type_JournalTag<🍆::Integer> {
public:
	🍆::WeakObject<Type_JournalTag> this_ptr;
	🍆::StrongObject<Type_JournalTag> get_this() {
		return 🍆::scast<Type_JournalTag>(this_ptr);
	}
	virtual ~Type_JournalTag() = default;

	Type_JournalTag(WeakObject<Type_JournalTag> this_ptr, 🍆::StrongObject<🍆::Integer>) : this_ptr(this_ptr) {
	}

	Type_JournalTag() = delete;

	Type_JournalTag(const Type_JournalTag&) = delete;

	Type_JournalTag(Type_JournalTag&&) = delete;

	void basefun_bring_back() {}

	virtual void privfun_bring_back() {
		(basefun_bring_back());
	}
	virtual void fun_bring_back() {
		(basefun_bring_back());
	}
	void basefun_t_(🍆::StrongObject<Type_JournalTag_<🍆::Integer>>, 🍆::StrongObject<🍆::Integer>) {}

	virtual void privfun_t_(🍆::StrongObject<Type_JournalTag_<🍆::Integer>> var_ttt, 🍆::StrongObject<🍆::Integer> var_tt) {
		(basefun_t_(var_ttt, var_tt));
	}
	virtual void fun_t_(🍆::StrongObject<Type_JournalTag_<🍆::Integer>> var_ttt, 🍆::StrongObject<🍆::Integer> var_tt) {
		(basefun_t_(var_ttt, var_tt));
	}
};

#define 🍆_Default_JournalTag
