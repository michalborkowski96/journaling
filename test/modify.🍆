import "../stdlib/vector.üçÜ"
import "../stdlib/journal_tag.üçÜ"

A {
	int v;
	constructor(int i){
		v = i;
	};
	fun int get() {
		return v;
	} noeffect;
	fun void set(int n) {
		v = n;
	} noeffect;
}

struct B {
	int v;
	constructor(int i){
		v = i;
	};
	fun void print(OutputStream o) {
		o.print_int(v);
	};
}

nojournal AMod : Modifier<A> {
	int diff;
	constructor(int d)(){
		diff = d;
	};
	fun void do(A arg) {
		arg.set(arg.get() + diff);
	};
	fun void undo(A arg) {
		arg.set(arg.get() - diff);
	};
}

nojournal BMod : Modifier<B> {
	int diff;
	constructor(int d)(){
		diff = d;
	};
	fun void do(B arg) {
		arg.v = arg.v + diff;
	};
	fun void undo(B arg) {
		arg.v = arg.v - diff;
	};
}

struct Main {
	strong OutputStream out;
	constructor(){
		out = new StdOut();
	};
	fun void main(){
		{
			var A a = new A(1);
			auto t = new JournalTag<A>(a);
			auto tt = new JournalTag<A>(a);
			tt.modify(new AMod(3));
			out.print_int_ln(a.get());
			t.bring_back();
			out.print_int_ln(a.get());
			tt.bring_back();
			out.print_int_ln(a.get());
		}
		out.print_newline();
		{
			var Vector<B> b = new Vector<B>();
			b.push_back(new B(1));
			b.push_back(new B(2));
			b.push_back(new B(-1));
			auto t = new JournalTag<Vector<B>>(b);
			auto bm = new BMod(5);
			for(auto i = b.begin(); i.neq(b.end()); i = i.next()) {
				b.modify(i, bm);
			}
			auto tt = new JournalTag<Vector<B>>(b);
			b.print(out);
			out.print_newline();
			t.bring_back();
			b.print(out);
			out.print_newline();
			tt.bring_back();
			b.print(out);
			out.print_newline();
		}
		out.print_newline();
		{
			var Vector<B> b = new Vector<B>();
			b.push_back(new B(1));
			b.push_back(new B(2));
			b.push_back(new B(-1));
			auto t = new JournalTag<Vector<B>>(b);
			b.modify_all(new BMod(5));
			auto tt = new JournalTag<Vector<B>>(b);
			b.print(out);
			out.print_newline();
			t.bring_back();
			b.print(out);
			out.print_newline();
			tt.bring_back();
			b.print(out);
			out.print_newline();
		}
	};
}
