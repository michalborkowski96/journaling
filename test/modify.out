4
1
4

{6, 7, 4}
{1, 2, -1}
{6, 7, 4}

{6, 7, 4}
{1, 2, -1}
{6, 7, 4}
